File: web_scrape/scrape.py
import requests
import bs4
import json

class JobInfo:
    def __init__(self, url=None, classification=None):
        self.url: str = url
        self.classification: list[tuple[str,list[str]]] = classification
    def get_tags(self):
        tags: list[str] = []
        for name, keywords in self.classification:
            tags.append(name)
        return tags

def classify(html: str):
    soup = bs4.BeautifulSoup(html, "html.parser")
    textElems: list[bs4.Tag] = soup.find_all(["h1", "h2", "h3", "h4", "h5", "p"])
    val = []
    categories = [
        ("software", ["software"]),
        ("web", ["web", "web-dev"]),
        ("frontend", ["frontend", "front-end", "front end"]),
        ("backend", ["backend", "back-end", "back end"]),
        ("embedded", ["embedded", "firmware"]),
        ("system", ["operating system"]),
        ("c++", ["c++"]),
        ("java", ["java"]),
        ("javascript", ["javascript", "typescript"]),
        ("c#", ["c#", ".net"])
    ]
    for name, keywords in categories:
        flag = False
        for elem in textElems:
            elemText = elem.decode_contents().strip()
            for i in range(0, len(elemText)):
                for keyword in keywords:
                    if i+len(keyword) <= len(elemText) and elemText[i : i+len(keyword)].lower() == keyword.lower():
                        val.append((name, keywords))
                        flag = True
                        break
                if flag: break
            if flag: break;
    return val

def scrape(targetUrls, domainMatch):
    def is_domainMatch(url: str):
        if url[0:4] != "http": return False
        for match in domainMatch:
            if url.find(match) != -1: return True
        return False
        
    subUrls: list[str] = []
    for url in targetUrls:
        req = requests.get(url)
        html = req.text
        print(html)
        soup = bs4.BeautifulSoup(html, "html.parser")
        links: list[bs4.Tag] = soup.find_all("a")
        for i, link in enumerate(links):
            subUrl: str = link.get("href")
            if subUrl and is_domainMatch(subUrl):
                print(f"{i+1}/{len(links)}: {subUrl}")
                subUrls.append(subUrl);
    print(f"{len(subUrls)} sites found")

    jobsFound: list[JobInfo] = []
    for i, url in enumerate(subUrls):
        print(f"Scraping {i+1}/{len(subUrls)}... ({url})")
        req = requests.get(url)
        html = req.text
        classification = classify(html)
        ji = JobInfo(url, classification)
        jobsFound.append(ji)
        
    for ji in jobsFound:
        print(ji.url, ji.get_tags())

def main():
    targetUrls = [
        "https://www.linkedin.com/jobs/search?keywords=software"
    ]
    domainMatch = [
        "www.linkedin.com/jobs"
    ]
    scrape(targetUrls, domainMatch)
    with open("data.json") as json_data:
        pass
        

if __name__ == "__main__":
    main()

File: web_scrape/scrape.ts
import axios, {AxiosResponse} from "axios";
import * as cheerio from "cheerio";
import * as fs from "fs";
import {JobData, JobInfo, tagCategories, DomainInfo} from "../include/types";
import * as lib from "../include/lib";
import { log } from "console";

const DIR = process.cwd();

function classify($: cheerio.CheerioAPI, domainInfo: DomainInfo | null): Array<string> {
    if (!domainInfo) {return [];} // stop compiler from complaining
    let desc = $(domainInfo.description);
    let val: Array<string> = [];
    let categories: Array<[string, Array<string>]> = tagCategories;
    for (let [name, keywords] of categories) {
        let flag = false;
        let descText = desc.text();
        for (let i = 0; i < descText.length; i++) {
            for (let keyword of keywords) {
                if (i+keyword.length <= descText.length
                && descText.substring(i, i+keyword.length).toLowerCase() == keyword.toLowerCase()) {
                    val.push(name);
                    flag = true;
                    break;
                }
            }
            if (flag) {break;}
        }
    }
    return val;
}

async function scrape(targetUrls: Array<string>, domainMatch: Array<DomainInfo>): Promise<Array<JobInfo>> {
    const FAST_SCRAPE = false;

    function get_domainMatch(url: string) {
        if (url.substring(0, 4) != "http") {return null;}
        for (let match of domainMatch) {
            if (url.includes(match.url)) {return match;}
        }
        return null;
    }

    const SCRAPE_LIMIT = 10000;
    let subUrls: Array<string> = [];
    for (let url of targetUrls) {
        let req = await axios.get(url);
        let html: string = req.data;
        const $ = cheerio.load(html);
        let linkElems = $("a[href]");
        for (let i = 0; i < linkElems.length; i++) {
            let subUrl = linkElems[i].attribs["href"].split("?")[0];
            if (get_domainMatch(subUrl) && subUrls.length < SCRAPE_LIMIT) {
                console.log(i+1+"/"+linkElems.length+": "+subUrl);
                subUrls.push(subUrl);
            }
        }
    }
    console.log(`${subUrls.length} sites found`);

    let jobsFound: Array<JobInfo> = [];
    let fastScrapeSetup: Array<Promise<AxiosResponse>> = [];
    let fastScrape: Array<AxiosResponse> = [];
    if (FAST_SCRAPE) {
        for (let i = 0; i < subUrls.length; i++) {
            let url = subUrls[i];
            fastScrapeSetup.push(axios.get(url));
        }
        log(`Parallel scraping (${fastScrapeSetup.length})...`);
        fastScrape = await Promise.all(fastScrapeSetup);
    }
    for (let i = 0; i < subUrls.length; i++) {
        let url = subUrls[i];
        console.log(`Scraping... ${url}`);
        try {
            let req = (FAST_SCRAPE? fastScrape[i] : await axios.get(url));
            let html: string = req.data;
            let $ = cheerio.load(html);
            let title = $("title").text();
            let match = get_domainMatch(url);
            let classification = classify($, match);
            let ji = new JobInfo(title, url, classification);
            if (match) { // stop compiler from complaining
                ji.posTitle = lib.prune_spaces($(match.title).text());
                ji.location = lib.prune_spaces($(match.location).first().text());
                ji.compensation = lib.prune_spaces($(match.compensation).text());
                // ji.description = lib.prune_spaces($(match.description).text()).substring(0, 20);
            }
            jobsFound.push(ji);
        } catch {
            console.log("Scrape failed.");
        }
    }
    for (let i = 0; i < jobsFound.length; i++) {
        let ji = jobsFound[i];
        console.log(`${i+1}/${jobsFound.length} ${ji.title} ${ji.url} [${ji.classification}]`);
    }
    return jobsFound;
}

async function main() {
    let targetUrls = [
        "https://www.linkedin.com/jobs/search?keywords=software",
        "https://www.linkedin.com/jobs/search?keywords=computer"
    ];
    let domainMatch: Array<DomainInfo> = JSON.parse(
        fs.readFileSync(DIR+"/scrape-domains.json", "utf-8")
    ).domains;
    try {
        let jobsFound = await scrape(targetUrls, domainMatch);
        let write = JSON.stringify(new JobData(jobsFound));
        fs.writeFileSync(DIR+"/../data/scrape.json", write, {flag: "w+"});
    } catch (e) {
        console.log("An error occurred:", e);
    }
}
main()